### ISSUE TITLE
Explore GitHub Actions

### ISSUE THREAD

**Issue by koenvervloesem** [rhasspy/rhasspy-services#5]
After [maxbachmann's comments](https://community.rhasspy.org/t/streamline-rhasspy-development-on-github/57/19) I'm exploring [GitHub Actions](https://github.com/features/actions) to see whether they are better suited than Travis CI:

https://github.com/rhasspy/rhasspy-speakers-cli-hermes/commit/36058e9fa1fb645a4d2ab99cd11313c34607cef6 (apparently I have commit rights now?)

I don't see any disadvantages yet. I propose to try Travis CI for a few repositories and GitHub Actions for a few others, to see how they compare, and then choose one.

**Comment by koenvervloesem** [comment]
So, I added GitHub Actions to:

- https://github.com/rhasspy/rhasspy-hermes
- https://github.com/rhasspy/rhasspy-microphone-cli-hermes
- https://github.com/rhasspy/rhasspy-speakers-cli-hermes
- https://github.com/rhasspy/rhasspy-nlu/pull/4

@maxbachmann is right: GitHub Actions definitely start up much faster. So while you're still waiting until your Travis CI worker is scheduled to run, your GitHub Actions are already running or have even finished.

Do you also receive emails with results from the GitHub Actions I have set up (passed/failed), @synesthesiam?

**Comment by maxbachmann** [comment]
I added a comment in #2 on how this could be implemented in a simple way

**Comment by koenvervloesem** [comment]
@synesthesiam: After having played with GitHub Actions for a few days, I definitely prefer it to Travis CI. It's fast, flexible, extensible, easy to grasp. We can always change later if needed, but I think it's best to focus our efforts now on GitHub Actions.

**Comment by maxbachmann** [comment]
@koenvervloesem when we have everything implemented it might be interesting to use a second ci solution that has ARM runners aswell

**Comment by koenvervloesem** [comment]
I created a [development status page](https://github.com/rhasspy/rhasspy/blob/master/development_status.md) in the rhasspy repository with a table with the status of tests, open issues and open pull requests for all repositories. This will give us a nice overview of how things in the ever expanding universe of Rhasspy services are :-D

**Comment by synesthesiam** [comment]
Very nice work on the development page! I think the number of services is finally stabilizing. I've got pretty good coverage of the various pieces of the original Rhasspy.

The next steps will be to get all the `Makefiles` updated with your various checks (again, thank you). @maxbachmann, do you know if it will be possible to build the ARM Docker images using GitHub Actions? I currently do this using qemu and Docker. In a perfect world, we'd have GitHub build the Debian packages and Docker images for `amd64`, `armv7l`, `arm64v8`, and `armv6l`.

**Comment by koenvervloesem** [comment]
I'm slowly going through all repositories, adding the GitHub Action, updating the Makefile, solving small issues, so when this is all done, the tests column of the development status page should be filled completely with all test statuses. This will also give us a better overview of where issues have to be solved.

**Comment by maxbachmann** [comment]
@koenvervloesem over at ProjectAlice I use Gitlab CI since I prefer it over Github Actions. Since we do have a repo for each skill we basically have exactly the same issue that we do not want to add the ci integration to each repo, since thats a lot of duplicated code that makes it harder to maintain.
I solved it using a template over there:
- https://github.com/project-alice-assistant/CI_Templates_Skills
so each skill simply adds this ci integration:
- https://github.com/project-alice-assistant/skill_TextInputWidget/blob/master/.gitlab-ci.yml
to run all these tests
https://gitlab.com/project-alice-assistant/skills/skill_DateDayTimeYear/pipelines/110337760
It is possible to overwrite everything derived from the included template, which makes this very well maintainable and since you can import from multiple files it is possible to split the ci into multiple files the way it is done with github actions as long as you import them all in a gitlab-ci.yml in the end ;)

I suppose with Github Actions the way to go would be to write actions like I did for the tagging

**Comment by synesthesiam** [comment]
I do like the ability to run the various checks, tests, etc. offline too though. I agree that it's a pain to keep all the Makefiles synced, and it cries out for a templating solution.

Git submodules aren't my favorite, but maybe one solution (locally at least, I think @maxbachmann is right for online CI) is to have a shared submodule repo with some `.mk` files we can include in our Makefiles.

**Comment by koenvervloesem** [comment]
Yes, i think that the short feedback loop of using the Makefile is important. I see the CI system just as an extension of that.

So for me this is not a question about GitHub Actions versus GitLab CI, but foremost about how to keep the Makefiles and supporting tools (e.g. the config files of isort, yamllint, pylint, and so on) maintainable.

I'm not familiar with Git submodules, would it make sense to put these configuration files also in a shared submodule repository?

**Comment by maxbachmann** [comment]
One thing to consider with submodules is that they always point to a specific commit sha. So when you update the Makefile you would need to go through all repositories and update this commit sha. This might be even easier to forget than updating the Makefile.

**Comment by koenvervloesem** [comment]
Good point, so either way there will be a maintenance burden...

Another solution is that I write a GitHub action like you did for the tagging, that just checks that the repository is using the latest Makefile and tool configurations (these can be compared with a template repository), and [creates a pull request](https://github.com/marketplace/actions/create-pull-request) when they don't match, so you can easily update this.

**Comment by maxbachmann** [comment]
yes that would be an option

**Comment by koenvervloesem** [comment]
Hi @synesthesiam I have around 2/3 of the repositories covered now with GitHub Actions being executed on each push or pull request. I noticed that many repositories have some small issues, such as new versions of dependencies not yet published on PyPI and some mypy errors. Could you take a look at them?

**Comment by synesthesiam** [comment]
I'm slowly making my way through. Didn't get as much time as I'd hoped this weekend; hopefully there will be more time available tomorrow.

I'd like to have a prototype system you and the other devs can try out by the end of the week :crossed_fingers:

**Comment by maxbachmann** [comment]
@koenvervloesem @synesthesiam I just added some sample code here: https://github.com/rhasspy/rhasspy-nlu-hermes/blob/docker/.github/workflows/tests.yml that uses `docker buildx` to crosscompile the docker containers for `amd64`, `armv7`, `arm64`, and `armv6`.
For some reason it fails to push the images to docker hub on the first try, but at least so far it always worked when simply retrying, which is why the ci integration simply runs the same command again when it fails the first time ;) (Maybe you know how to fix this properly)

**Comment by maxbachmann** [comment]
this only cross compiles the docker stuff. Pyinstaller apparently needs to be cross compiled aswell

**Comment by maxbachmann** [comment]
I am currently try to cross compile using pyinstaller aswell, but it still requires to compile the bootloader aswell since there are no precompiled precompiled bootloaders for arm

**Comment by synesthesiam** [comment]
It would be simpler not to use PyInstaller, yeah. Some of the services need dev tools to install pip packages, though, like snowboy or pocketsphinx. This made the Docker images much bigger than necessary, unless I started playing tricks with a virtual environment in a separate build stage.

Since I already had the PyInstaller scripts for building the Debian packages, I just took the lazy route :) But this may not be the best way for cross-compiling on GitHub, so I'm open to other ideas.

**Comment by maxbachmann** [comment]
I just compiled it wihtout pyinstaller in docker and it did not really get much bigger (59mb vs 64mb) but apparently runs a lot faster (probably could be optimised by selecting a different base docker image.
@synesthesiam I never used pyinstaller before. Is there any other advantage in using pyinstaller beside the packaging (which is not required in the docker container)?

**Comment by synesthesiam** [comment]
No other advantage for me. I'd also rather distribute the source code than binary blobs, so not using PyInstaller is a win there.

**Comment by maxbachmann** [comment]
Can you add a secret to the repos using docker that is named `DOCKER_PASSWORD` and holds the password for docker hub so I can deploy the images?

**Comment by maxbachmann** [comment]
https://hub.docker.com/r/maxbachmann/example/tags is the cross compiled docker container with lite-buster and alpine

**Comment by synesthesiam** [comment]
> Can you add a secret to the repos using docker that is named `DOCKER_PASSWORD` and holds the password for docker hub so I can deploy the images?

How exactly do you mean? There was a particular process for Travis CI to do this. Is there something specific for Github Actions or Docker itself?

**Comment by synesthesiam** [comment]
> https://hub.docker.com/r/maxbachmann/example/tags is the cross compiled docker container with lite-buster and alpine

This is impressive, and is going to save so much time when deploying new versions. Thank you!

**Comment by maxbachmann** [comment]
@synesthesiam in the rebot click on `settings > Secrets > Add a new secret` then you can add the name and value of the secret which will be used for github actions

**Comment by synesthesiam** [comment]
OK, I've added it to rhasspy-nlu-hermes. If that works right for you, I'll go ahead and add it to the other services.

**Comment by maxbachmann** [comment]
Can you upload rhasspy-hermes==0.1.6 to pypi? Otherwise it will never run since it can not install the correct dependencies: https://github.com/rhasspy/rhasspy-nlu-hermes/issues/2

**Comment by maxbachmann** [comment]
@synesthesiam are you sure you added it?
I did just rerun the github action and it failed: https://github.com/rhasspy/rhasspy-nlu-hermes/runs/430246559?check_suite_focus=true
The error sounds like the variable `DOCKER_PASSWORD` was empty

**Comment by synesthesiam** [comment]
Got it to work by adding the `env` section in the YML to expose the secret to the workflow.

**Comment by maxbachmann** [comment]
hm makes sence that this is required ... Not sure how this worked in my fork before.
Thanks for fixing
